### DISCUSSION TITLE
Traffic_shifting not working with ISTIO 1.8.1 and Kiali 1.22 or 1.26

### DISCUSSION THREAD

**Discussion by sabikm9876** [kiali/kiali#3640] [Q&A]
We have been trying to figure issues with our ISTIO where we are trying to use the 'traffic shifting' in kiali. e used to use 'updated_weights' but it looks like that is no longer supported. So we are now in Istio 1.8.1 and kiali 22. When we deploy and look in kiali, traffic shifting, request routing are all greyed out and so are the other options. We tried updating the weights through the yaml file directly but it was still not reflecting the correct weights as we run the load tests. Can you help explain what we are missing and how do we fix this?

**Comment by israel-hdez** [comment]
Can you, please, tell how are you deploying Kiali? (via the kiali-operator, via helm charts, something else?)

Also, you will need to post your YAML files to see where the issue may be (you can mask any sensitive data).

**Comment by lucasponce** [comment]
@sabikm9876 I think that you are generating the files using kubectl or any CLI and then you try to see them on Kiali, isn't it?

When Istio config is generated from a CLI, all artifacts are visible on the "Istio Config" section, where you can see and edit them through the YAML editor.

Kiali offers under the Service a set of Wizards that managed "a complete scenario", I mean, a Kiali Wizard creates in a high level view all VirtualService, DestinationRule, Gateway and PeerRequest objects.

When you have created a Wizard from the scratch, you'll see that the whole scenario can be updated, if there were some previous config created under a service, Kiali allows to delete them but only updates those Istio config previously generated.

(You will see that Istio configuration generated by Kiali is labeled (and annotated) with 'kiali_wizard' label).

In short:
- For easy scenarios, try to delete and recreate the Istio config through the Kiali Wizard.
- Istio Config section will allow to edit them through the editor.
- Existing Istio config created via CLI could be visualized in Kiali if VS/DR are labeled with 'kiali_wizard=<wizard_type>' (kiali_wizard=request_routing, or kiali_wizard=traffic_shifting), but the support of external Istio files through wizard is still experimental as Kiali doesn't cover yet some cornercases and complex configurations (Wizard may miss some data).

Let me know if this adds more info, or feel free to comment any other question you may have around it.

**Reply by sabikm9876** [reply]
Hi Edgar and Lucas, Thank You both for replying to my inquiry here in Git hub kiali discussion. First of all, you are correct, I am generating the files using kubectl and deploying my containers/pods to istio. I am seeing the 'istio config'in Kiali and the yaml files which corresponds to my Yaml files. I also use the 'label: kiali_wizard : traffic_shifting' or 'kiali_wizard_request_routing'. And I do update that instead of using 'request_routing' or 'traffic_shifting' in kiali. I do this as my traffic load is running. I notice that it  stops the traffic  and the shift  to the other version does not work.. Since we will use this for canary deployment we would like to be able to do this fast and not have to go through the Yaml files to update it. Not very clear on the 'When you have created a Wizard from the scratch, you'll see that the whole scenario can be updated, if there were some previous config created under a service, Kiali allows to delete them but only updates those Istio config previously generated.'. How do you create a Wizard from 'scratch'? I would like it that everything we define in Yaml would redlect in the 'traffic_shifting' or 'request_routing'. Is there anyway to load and deploy the Yaml files through the wizard in kiali?  Also, I was trying to select both 'traffic_shifting' and 'request_routing' in the 'label: kiali_wizard' but have failed to no avail.  Thank You again. All these explains what is what I have observed in kiali so far. Also, any useful sample scripts would also be great that would help with our canary deployments. We are using 3 versions. With use of URI prefixes and headers. I will attach the file soon.

**Reply by lucasponce** [reply]
Thanks @sabikm9876.
The Kiali Wizards are designed to use them from high level perspective.

I recommend to take a look to the tutorial https://kiali.io/documentation/latest/tutorial/ to the "Connect" section, there are some examples of how to use the tutorial.

If you want to keep creating the Istio config through kubectl and you'd like to open them as well from the wizard, then those VS/DR should have the Kiali labels that I've commented.

Note that this last scenario it may be experimental and perhaps some particular config you have is not yet fully supported. I'd be happy to take a look at them if you could provide some demo config similar as you are using (without no names in case there is any sensitive data, if you prepare an example about what is not working, using a demo - bookinfo or similar - that would help us in the same way).

**Comment by sabikm9876** [comment]
Hi, Thank you for the link for the examples. I will take a look. Attached is our yaml files with all sensitive info stripped off or cut off. Thank You.apiVersion: v1
kind: Service
metadata:
  name: deletepatientotcherbalcodes-dep
  labels:
    app: deletepatientotcherbalcodes-dep
spec:
  ports:
  - port: 8951
    name: http
  selector:
    app: deletepatientotcherbalcodes-dep
  type: ClusterIP
---
apiVersion: v1
kind: ServiceAccount
metadata:
  name: deletepatientotcherbalcodes-dep
  labels:
    account: deletepatientotcherbalcodes-dep
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: deletepatientotcherbalcodes-dep-v1
  labels:
    app: deletepatientotcherbalcodes-dep
    version: v1
spec:
  replicas: 1
  selector:
    matchLabels:
      app: deletepatientotcherbalcodes-dep
      version: v1
  template:
    metadata:
      labels:
        app: deletepatientotcherbalcodes-dep
        version: v1
    spec:
      serviceAccountName: deletepatientotcherbalcodes-dep
      containers:
      - name: deletepatientotcherbalcodes-dep
        image: image/deletepatientotcherbalcodes:
        imagePullPolicy: IfNotPresent
        ports:
        - containerPort: 8951
---
apiVersion: v1
kind: ServiceAccount
metadata:
  name: deletepatientotcherbalcodes-dep
  labels:
    account: deletepatientotcherbalcodes-dep
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: deletepatientotcherbalcodes-dep-v2
  labels:
    app: deletepatientotcherbalcodes-dep
    version: v2
spec:
  replicas: 1
  selector:
    matchLabels:
      app: deletepatientotcherbalcodes-dep
      version: v2
  template:
    metadata:
      labels:
        app: deletepatientotcherbalcodes-dep
        version: v2
    spec:
      serviceAccountName: deletepatientotcherbalcodes-dep
      containers:
      - name: deletepatientotcherbalcodes-dep
        image: image/deletepatientotcherbalcodes:
        imagePullPolicy: IfNotPresent
        ports:
        - containerPort: 8951

---
apiVersion: v1
kind: ServiceAccount
metadata:
  name: deletepatientotcherbalcodes-dep
  labels:
    account: deletepatientotcherbalcodes-dep
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: deletepatientotcherbalcodes-dep-v3
  labels:
    app: deletepatientotcherbalcodes-dep
    version: v3
spec:
  replicas: 1
  selector:
    matchLabels:
      app: deletepatientotcherbalcodes-dep
      version: v3
  template:
    metadata:
      labels:
        app: deletepatientotcherbalcodes-dep
        version: v3
    spec:
      serviceAccountName: deletepatientotcherbalcodes-dep
      containers:
      - name: deletepatientotcherbalcodes-dep
        image: image/deletepatientotcherbalcodes:
        imagePullPolicy: IfNotPresent
        ports:
        - containerPort: 8951


apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: deletepatientotcherbalcodes
  labels:
    kiali_wizard: traffic_shifting
spec:
  hosts:
  - "*"
  gateways:
  - pcq-test-repl-istio-gateway
  http:
    - match:
      - headers:
          environment:
            exact: test
        uri: 
            prefix: /deletepatientotcherbalcodes
      route:
      - destination:
          host: deletepatientotcherbalcodes-dep.
          subset: v2
          port:
                number: 8951
        weight: 30
    - match:
      - uri: 
            prefix: /deletepatientotcherbalcodes
      route:
      - destination:
          host: deletepatientotcherbalcodes-dep.
          subset: v1
          port:
                number: 8951
        weight: 70
    - match:
      - uri: 
            prefix: /deletepatientotcherbalcodes
      route:
      - destination:
          host: deletepatientotcherbalcodes-dep.
          subset: v3
          port:
                number: 8951
        weight: 0
---

apiVersion: networking.istio.io/v1alpha3
kind: DestinationRule
metadata:
  name: deletepatientotcherbalcodes
  labels:
    kiali_wizard: traffic_shifting
spec:
  host: deletepatientotcherbalcodes-dep.
  subsets:
  - name: v1
    labels:
      version: v1
  - name: v2
    labels:
      version: v2
  - name: v3
    labels:
      version: v3

**Reply by lucasponce** [reply]
Thanks, perhaps could you updload them with a gist https://gist.github.com/ ? that would help to read/download them better

**Reply by jmazzitelli** [reply]
You can supply yaml properly formated if you just wrap it here in back-tick marks. Example:
\`\`\`yaml
yourYaml: here
\`\`\`

I edited your post with those markers so your yaml now looks correctly displayed in your previous comment.

**Comment by sabikm9876** [comment]
Hi Lucas, Here is the link > https://gist.github.com/maryjsyamlfile

**Reply by lucasponce** [reply]
I can't open the link, perhaps it's not a public link ?

**Reply by jmazzitelli** [reply]
@lucasponce see above - I formated the yaml properly in the previous comment now

**Reply by lucasponce** [reply]
Well, I think it's easier and handy if she could just upload a public gist from her account
https://gist.github.com/sabikm9876

Then we can direcltly download the files.

@sabikm9876 take a look to my account as an example https://gist.github.com/lucasponce

**Reply by lucasponce** [reply]
In yaml world it's handy to get the raw files in case that a label is not working due a bad space or similar kind of problem (that was my motivation).

**Reply by lucasponce** [reply]
Also, in the comments, the yaml posted seemed in that way, that there may be some typo/space in the yaml that is preventing to be read correctly. @sabikm9876 if you can upload them in your gist account as a public gits I can see. I'll take a look locally on them.

**Comment by sabikm9876** [comment]
I may have not uploaded it earlier. I just uploaded as a public gist.

**Reply by lucasponce** [reply]
I'm afraid I can't see the gists:
![image](https://user-images.githubusercontent.com/1662329/106251450-46682e00-6215-11eb-9300-313778747ea1.png)

But I guess I can build an example for your scenario, I'll send a followup comment with captures.

**Comment by lucasponce** [comment]
I've run a demo using Istio 1.8.2 and I think it may resolve the question you may have.

So, I'll present several steps (perhaps some may be redundant) in the spirit to clarity and avoid misunderstandings about the scope of the problem.

Step 1: I've installed a new kubernetes cluster with a pretty standard Istio 1.8.2 setup [1].

Step 2: I've upgraded Kiali to latest version v1.29.0 following similar step as described in [2].

Step 3: Opening Kiali I can verify my version:
![image](https://user-images.githubusercontent.com/1662329/106251935-eaea7000-6215-11eb-8c6c-1f24065d00c7.png)

Step 4: Similar as you described, I've created some Istio configuration using kubectl, not using Kiali. In my case, and I've used the Istio examples [3] but looking into your yaml, I think it will help us for the same question. As a result, all these new Istio config files can be seen within Kiali -> Istio Config page:
![image](https://user-images.githubusercontent.com/1662329/106252213-4e749d80-6216-11eb-82fe-65d4685f577b.png)

Step 5: This configuration hasn't been generated from Kiali itself, so when we navigate to a particular service, like "reviews" on this example, we see that Kiali doesn't enable the wizard, we can see the Istio config related to the service, but the only action it allows us is to delete them. That's expected.
![image](https://user-images.githubusercontent.com/1662329/106252432-8f6cb200-6216-11eb-85c2-c5421851b23c.png)

The reason why this is expected is that all combinations of configuration are not fully supported from the Wizard yet, but a user can generate a high level scenario from Kiali.

Step 6: As a EXPERIMENTAL step, if you label the pair of VirtualService and DestinationRule used, you can enable them into Kiali Wizard. Note that Kiali always allows to edit Istio config from the editor as well. Indeed, I used it to add these labels.
![image](https://user-images.githubusercontent.com/1662329/106252699-d195f380-6216-11eb-86d0-0929a4e8884d.png)

Step 7: If you go back to the "reviews" service, you will see that Kiali detects that the Istio config is editable by the wizard.
![image](https://user-images.githubusercontent.com/1662329/106252851-043fec00-6217-11eb-8e92-f87f8bf6e99b.png)

Note, on this case I used the pair kiali_wizard: traffic_shifting, but it can be used other label value as "request_routing", etc.
I insist, this is experimental, it's part of the implementation detail, just a workaround for users that want to enable external Istio configuration into the Kiali Wizard.

Step 8: First time you open the Traffic Shifting Wizard, probably you will see something like this:
![image](https://user-images.githubusercontent.com/1662329/106253086-54b74980-6217-11eb-9328-41714762caf2.png)

It's not a problem, as commented, the Wizard tries to link workloads with VirtualService destination and if something is not 100% correct it may no recognize it.

But you can update the config with the existing workloads, for example, I've updated my config like this:
![image](https://user-images.githubusercontent.com/1662329/106253207-7c0e1680-6217-11eb-983a-85c6d882e559.png)

Step 9: If we examine the Istio config, we will see how the Wizard manages the values, allowing the user to modify them from a high level scenario:
![image](https://user-images.githubusercontent.com/1662329/106253248-892b0580-6217-11eb-9da8-dcb466360f28.png)

Conclusion.

- If a Service has a existing configuration, you can always manage it through the YAML editor.
- If you want to use the wizard, a recommended way is to delete the pair of VS/DR ("Delete Traffic Routing" action) and create a new scenario from the Wizard, you will see how Kiali generates the proper config for you.
- If you want to manage existing configuration from the wizard, as an experimental step, at least the pair of files VS/DR needs a proper "kiali_wizard" labeling as described on this comment. Note that Wizards don't support all potential combinations of configurations and this is an active development area in the project, but any feedback from users on this area is always welcome.

Hope this comment helps to progress in your case and at least it helps you to understand why you didn't see the Istio config available in the Kiali Wizard before.

Feel free to add any question / comment around.

Thanks,
Lucas


[1] https://istio.io/latest/docs/setup/getting-started/
[2] https://kiali.io/documentation/v1.29/tutorial/#03-update-kiali
[3] https://istio.io/latest/docs/tasks/traffic-management/traffic-shifting/

**Comment by sabikm9876** [comment]
Hi Lucas,   Thank you very much for the detailed demo screen prints . I am already doing all of this and the 'kiali_wizard: traffic_shifting' was in the VS/DR yaml files. In Kiali, when I go into reviews/service, I then change the weights around. I go back into 'istio_config'  and I do see that the weights are reflected but all of the 'match/uri/prefix' and 'match/uri/prefix/header' are all gone, This then cause all of the traffic to stop as it doesn't know where to send the traffic to. I am not sure whether if we have the request_routing added that it will retain this information even if we change the weights. But we can only add one option at a time. Or is there a way and will it help? Again Thank you for looking into this.

**Reply by lucasponce** [reply]
Yes, the "Traffic Shifting" Wizard is a simplified scenario where there is only one matching rule.
If for existing VS/DR config you use the labels:

```
kiali_wizard: request_routing
```

You'll tell Kiali to try to open it from the Request Routing Wizard. 
This Wizard is more complex and allow to define one to many rules.

But again, it's probably that you'll find some use case that it's not yet supported, so if that's the case, you'll need to use the YAML editor to manage the config.

In any case, I'm interested to get your feedback/experience, as commented, to allow manage existing Istio config is a complex scenario where we are working on, so any real use cases will help us to drive this effort (this effort will be tracked in https://github.com/kiali/kiali/issues/3652 issue).

**Comment by lucasponce** [comment]
@sabikm9876 Did you try to use the request_routing wizard ?
At least, did i help in your scenario ?

Thanks
